packages/runtime/datastore/src/dataStoreRuntime.ts

```diff
@@ -650,66 +652,166 @@ export class FluidDataStoreRuntime
+				};
+
+				const remoteChannelContext = this.createRemoteChannelContext(
+					attachMessage,
+					summarizerNodeParams,
+				);
+				this.contexts.set(id, remoteChannelContext);
+			}
+		}
+	}
+
+	/**
+	 * Process messages for this data store. The messages here are contiguous messages in a batch.
+	 * @param message - The core message with common properties for all the messages.
+	 * @param messageContents - The contents, local metadata and clientSequenceNumbers of the messages.
+	 * @param local - true if the messages were originally generated by the client receiving it.
+	 */
+	public processMessages(
+		message: ISequencedRuntimeMessageCore,
+		messageContents: IRuntimeMessageContents[],
+		local: boolean,
+	): void {
+		this.verifyNotClosed();
+
 		try {
-			// catches as data processing error whether or not they come from async pending queues
 			switch (message.type) {
-				case DataStoreMessageType.Attach: {
-					const attachMessage = message.contents as IAttachMessage;
-					const id = attachMessage.id;
-
-					// We need to process the GC Data for both local and remote attach messages
-					processAttachMessageGCData(attachMessage.snapshot, (nodeId, toPath) => {
-						// Note: nodeId will be "/" unless and until we support sub-DDS GC Nodes
-						const fromPath = `/${this.id}/${id}${nodeId === "/" ? "" : nodeId}`;
-						this.dataStoreContext.addedGCOutboundRoute(fromPath, toPath, message.timestamp);
-					});
-
-					// If a non-local operation then go and create the object
-					// Otherwise mark it as officially attached.
-					if (local) {
-						assert(
-							this.pendingAttach.delete(id),
-							0x17c /* "Unexpected attach (local) channel OP" */,
-						);
-					} else {
-						assert(!this.contexts.has(id), 0x17d /* "Unexpected attach channel OP" */);
-
-						const summarizerNodeParams = {
-							type: CreateSummarizerNodeSource.FromAttach,
-							sequenceNumber: message.sequenceNumber,
-							snapshot: attachMessage.snapshot,
-						};
-
-						const remoteChannelContext = this.createRemoteChannelContext(
-							attachMessage,
-							summarizerNodeParams,
-						);
-						this.contexts.set(id, remoteChannelContext);
-					}
-					break;
-				}
-
 				case DataStoreMessageType.ChannelOp:
-					this.processChannelOp(message, local, localOpMetadata);
+					this.processChannelMessages(message, messageContents, local);
+					break;
+				case DataStoreMessageType.Attach:
+					this.processAttachMessages(message, messageContents, local);
 					break;
 				default:
 			}
-
-			this.emit("op", message);
 		} catch (error) {
 			throw DataProcessingError.wrapIfUnrecognized(
 				error,
 				"fluidDataStoreRuntimeFailedToProcessMessage",
 				message,
 			);
 		}
+
+		for (const { contents } of messageContents) {
+			this.emit("op", { ...message, contents });
+		}
+	}
+
+	/**
+	 * This is still here for back-compat purposes because it exists on IFluidDataStoreChannel. Once it is removed from
```

Comment:
Please use // back-compat comments to easier find them later

Reply:
Done.